bus/uploadingsectors.go

```diff
@@ -0,0 +1,111 @@
+package bus
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+)
+
+const (
+	// cacheExpiry is the amount of time after which an upload is pruned from
+	// the cache, since the workers are expected to finish their uploads this is
+	// there to prevent leaking memory, which is why it's set at 24h
+	cacheExpiry = 24 * time.Hour
+)
+
+type (
+	uploadingSectorsCache struct {
+		mu      sync.Mutex
+		uploads map[api.UploadID]*ongoingUpload
+	}
+
+	ongoingUpload struct {
+		mu              sync.Mutex
+		started         time.Time
+		contractSectors map[types.FileContractID][]types.Hash256
+	}
+)
+
+func newUploadingSectorsCache() *uploadingSectorsCache {
+	return &uploadingSectorsCache{
+		uploads: make(map[api.UploadID]*ongoingUpload),
+	}
+}
+
+func (ou *ongoingUpload) addSector(fcid types.FileContractID, root types.Hash256) {
+	ou.mu.Lock()
+	defer ou.mu.Unlock()
+	ou.contractSectors[fcid] = append(ou.contractSectors[fcid], root)
+}
+
+func (ou *ongoingUpload) sectors(fcid types.FileContractID) (roots []types.Hash256) {
+	ou.mu.Lock()
+	defer ou.mu.Unlock()
+	if sectors, exists := ou.contractSectors[fcid]; exists && time.Since(ou.started) < cacheExpiry {
+		roots = append(roots, sectors...)
+	}
+	return
+}
+
+func (usc *uploadingSectorsCache) addUploadingSector(uID api.UploadID, fcid types.FileContractID, root types.Hash256) error {
+	defer usc.pruneExpiredUploads()
```

Comment:
(nit) would be nice if we could make use of the fact that we already acquire the lock here instead of acquiring it, adding a sector, releasing it, acquiring it, pruning and releasing again.

Also why prune after adding an upload instead of after deleting one? Doesn't matter but finishing seems closer related to pruning than adding.

Reply:
I originally had `add` and `remove` sector and pruned on both occasions. I'll move it to `finish`